# Python de cero a experto
**Autor:** Luis Miguel de la Cruz Salas

<a href="https://github.com/luiggix/Python_cero_a_experto">Python de cero a experto</a> by Luis M. de la Cruz Salas is licensed under <a href="https://creativecommons.org/licenses/by-nc-nd/4.0?ref=chooser-v1">Attribution-NonCommercial-NoDerivatives 4.0 International</a>


**Objetivos.**
...

## El huevo cocido perfecto

¿Cuál es el tiempo ideal para cocinar un huevo?

<img src="./Figuras/Huevo_cocido.jpg"  style="width: 300px;"/>


Para lograr un **huevo cocido** suave, la clara debe haberse calentado el tiempo suficiente para coagular a una temperatura superior a 63 $^o$C, pero la yema no debe calentarse por encima de 70 $^o$C.

Para lograr un **huevo duro**, el centro de la yema debe de alcanzar los 70 $^o$C.

### Fórmula para calcular el tiempo de cocción.
La siguiente fórmula expresa el tiempo $t$, en segundos, que le toma a la yema alcanzar la temperatura $T_y$, en grados Celsius.

$$t = \dfrac{M^{2/3} c \rho^{1/3}}{K \pi^2 (4\pi/3)^{2/3}} \ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right]$$

donde las propiedades son:

- $M$ masa; 
- $\rho$ densidad; 
- $c$ capacidad calorífica específica; 
- $K$ conductividad térmica; 
- $T_w$ es la temperatura de ebullición del agua;
- $T_o$ es la temperatura original del huevo antes de meterlo al agua;
- $T_y$ es la temperatura que debe alcanzar la yema.

### Ejercicio 1.
Calcular el tiempo de cocción necesario para un huevo duro pequeño de 47 g 
(la masa de un huevo grande es 67 g), para cuando la temperatura inicial 
del huevo es:
1. Temperatura ambiente: $T_o$ = 20 $^o$C.
2. Temperatura en el refrigerador: $T_o$ = 4 $^o$C.

**Datos**:
- $M$ = 47 g 
- $\rho$ = 1.038 g / cm$^3$
- $c$ = 3.7 J / g K
- $K$ = 5.4 $\times 10^{-3}$ W / cm K
- $T_w$ = 100 $^o$C
- $T_y$ = 70 $^o$C

**<font color="#126534">SOLUCIÓN.</font>**<br>
Primera parte de la fórmula: $M^{2/3} c \rho^{1/3}$

In [38]:
# Definir los datos iniciales
M   = 47    # Un entero
rho = 1.038 # Un flotante
c   = 3.7   # Otro flotante

print(type(M), type(rho), type(c))
print(id(M), id(rho), id(c))
print(M, rho, c)

<class 'int'> <class 'float'> <class 'float'>
93857122751232 139885541568848 139885541569328
47 1.038 3.7


In [39]:
numerador = M**(2/3) * c * rho**(1/3)
numerador

48.790216719661984

**Revisar**: <a href="./T01_Etiquetas_y_Palabras_Reservadas.ipynb">Etiquetas y palabras reservadas </a>

Podríamos usar caracteres matemáticos:

In [40]:
print(chr(0x1D70C))

𝜌


In [41]:
𝜌 = 1.038

In [42]:
M**(2/3) * c * 𝜌**(1/3)

48.790216719661984

**Revisar**: <a href="./T02_Expr_Decla_Tipos_Oper.ipynb">Expresiones, Declaraciones, Tipos y Operadores</a>

Segunda parte de la fórmula: $K \pi^2 (4\pi/3)^{2/3}$

In [43]:
# Denominador, necesitamos el valor de Pi.
import math
# Se puede importar la biblioteca math y ponerle otro nombre 
#import math as m 
# Se puede importat solo una función de la biblioteca y ponerle un nombre
#from math import sinh as senoHiperbolico

In [44]:
K = 5.4e-3
denominador = K * math.pi**2 * (4 * math.pi / 3)**(2/3)

In [45]:
print(denominador)

0.13849026450902358


In [46]:
chr(0x03C0)

'π'

In [47]:
π = math.pi
K * π**2 * (4 * π / 3)**(2/3)

0.13849026450902358

Tercera parte de la fórmula: $\ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right]$

**Caso 1.**<br>
$T_o$ = 20 $^o$C.

In [48]:
To = 20  # Temp. ambiente
Tw = 100
Ty = 70

In [49]:
logaritmo = math.log(0.76 * (To - Tw) / (Ty - Tw))

Fórmula completa:

In [50]:
t = numerador / denominador * logaritmo
print(t)

248.86253747844736


**Caso 2.**<br>
$T_o$ = 4 $^o$C.

In [51]:
To = 4  # Temp. ambiente
Tw = 100
Ty = 70

In [52]:
logaritmo = math.log(0.76 * (To - Tw) / (Ty - Tw))

In [53]:
t = numerador / denominador * logaritmo
print(t)

313.09454902221637


### Ejercicio 2.
Imprimir el tiempo de los casos del ejercicio 1 en el siguiente formato usando *cadenas*:<br>
 `El tiempo de cocción óptimo es: 313.1 [s] (5.2 [m])`

In [72]:
To = 4
logaritmo = math.log(0.76 * (To - Tw) / (Ty - Tw))
t1 = numerador /denominador * logaritmo

To = 20
logaritmo = math.log(0.76 * (To - Tw) / (Ty - Tw))
t2 = numerador /denominador * logaritmo

In [77]:
print('El tiempo de cocción óptimo es:', t1,' [s] (', t1/60, ' [m])')
print('El tiempo de cocción óptimo es:', t2,' [s] (', t2/60, ' [m])')

El tiempo de cocción óptimo es: 313.09454902221637  [s] ( 5.218242483703606  [m])
El tiempo de cocción óptimo es: 248.86253747844736  [s] ( 4.147708957974123  [m])


In [78]:
cadena1 = 'El tiempo de cocción óptimo es: '
cadena2 = str(t1) + ' [s] (' + str(t1) + ' [m])'
cadena3 = str(t2) + ' [s] (' + str(t2) + ' [m])'
print(cadena1 + cadena2)
print(cadena1 + cadena3)

El tiempo de cocción óptimo es: 313.09454902221637 [s] (313.09454902221637 [m])
El tiempo de cocción óptimo es: 248.86253747844736 [s] (248.86253747844736 [m])


In [80]:
cadena2 = '{} [s] ( {} [m])'.format(t1, t1/60)
cadena3 = '{} [s] ( {} [m])'.format(t2, t2/60)
print(cadena1 + cadena2)
print(cadena1 + cadena3)

El tiempo de cocción óptimo es: 313.09454902221637 [s] ( 5.218242483703606 [m])
El tiempo de cocción óptimo es: 248.86253747844736 [s] ( 4.147708957974123 [m])


In [83]:
cadena2 = '{:.1f} [s] ( {:.1f} [m])'.format(t1, t1/60)
cadena3 = '{:.1f} [s] ( {:.1f} [m])'.format(t2, t2/60)
print(cadena1 + cadena2)
print(cadena1 + cadena3)

El tiempo de cocción óptimo es: 313.1 [s] ( 5.2 [m])
El tiempo de cocción óptimo es: 248.9 [s] ( 4.1 [m])


In [84]:
print('El tiempo de cocción óptimo es: {:0.1f} [s] ({:0.1f} [m])'.format(t1, t1/60))
print('El tiempo de cocción óptimo es: {:0.1f} [s] ({:0.1f} [m])'.format(t2, t2/60))

El tiempo de cocción óptimo es: 313.1 [s] (5.2 [m])
El tiempo de cocción óptimo es: 248.9 [s] (4.1 [m])


### Ejercicio 3.
Hacer una lista de tiempos de cocción para temperaturas de huevos, desde la que se tiene en el refrigerador hasta temperatura ambiente, en pasos de 1$^o$C.
- Imprimir dos columnas: temperatura y tiempo.
- Imprimir los tiempos en minutos y segundos.
- Imprimir un encabezado para identificar las columnas.

<a href="./T03_Estructura_de_Datos.ipynb">Estructura de datos</a>

<a href="./T04_Control_de_flujo.ipynb">Control de flujo</a>

**<font color="#126534">SOLUCIÓN.</font>**<br>
Recordemos la fórmula $$t = \dfrac{M^{2/3} c \rho^{1/3}}{K \pi^2 (4\pi/3)^{2/3}} \ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right]$$

`numerador` = $M^{2/3} c \rho^{1/3}$

`denominador` = $K \pi^2 (4\pi/3)^{2/3}$

`logaritmo` = $ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right]$

Observamos que lo único que cambia es `logaritmo`.

In [68]:
for To in range(4, 20):
    print(To)

4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19


In [69]:
for To in range(4, 21):
    logaritmo = math.log(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * logaritmo
    print(t)

313.09454902221637
309.4055027800624
305.67741828677
301.9094604759048
298.100767196758
294.2504480302776
290.3575830395756
286.42122145062837
282.44038025843554
278.4140427535251
274.34115696328047
270.22063400211084
266.05134632399177
261.83212587036
257.5617621057524
253.23899993292162
248.86253747844736


In [70]:
for To in range(4, 21):
    logaritmo = math.log(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * logaritmo
    print(To, t/60)

4 5.218242483703606
5 5.156758379667707
6 5.094623638112833
7 5.03182434126508
8 4.968346119945966
9 4.90417413383796
10 4.8392930506595935
11 4.773687024177139
12 4.7073396709739255
13 4.640234045892085
14 4.572352616054674
15 4.503677233368514
16 4.434189105399863
17 4.363868764506
18 4.292696035095873
19 4.220649998882027
20 4.147708957974123


In [71]:
print('Temperatura \t Tiempo')
for To in range(4, 21):
    logaritmo = math.log(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * logaritmo
    t_min = int(t / 60)
    t_seg = int(t - t_min * 60) 
    print(' {} \t\t {}:{}'.format(To, t_min, t_seg))

Temperatura 	 Tiempo
 4 		 5:13
 5 		 5:9
 6 		 5:5
 7 		 5:1
 8 		 4:58
 9 		 4:54
 10 		 4:50
 11 		 4:46
 12 		 4:42
 13 		 4:38
 14 		 4:34
 15 		 4:30
 16 		 4:26
 17 		 4:21
 18 		 4:17
 19 		 4:13
 20 		 4:8


In [ ]:
Ts = []
ts = []
for i in range(4,21):
    Ts.append(i)
    t = (num1 /den1) * math.log(0.76 * (i - Tw) / (Ty - Tw))
    ts.append(t/60)
print(Ts)
print(ts)

### Ejercicio 5:
De la lista creada anteriormente determinar que temperatura debería tener el huevo originalmente para que el tiempo de cocción sea de 5 minutos.

In [ ]:
# Encontrar el intervalo donde cae el valor de 5 minutos
Ip1 = 0 # índice donde se pasa de 5 a menos de 5
for i,v in enumerate(ts):
    if v < 5.0:
        Ip1 = i
        print('Tiempos [s]: ts[{}] : {:2.5f}; ts[{}] : {:2.5f}'.format(i-1,ts[i-1],i,v))
        print('Temps   [C]: Ts[{}] : {:2.5f}; Ts[{}] : {:2.5f}'.format(i-1,Ts[i-1],i,Ts[i]))
        break

# Calcular la pendiente:
m = (Ts[Ip1] - Ts[Ip1-1]) / (ts[Ip1] - ts[Ip1-1])
# Calcular la ordenada al origen:
b = Ts[Ip1] - m * ts[Ip1]
print('Pendiente : {:2.5f}'.format(m))
print('Ordenada  : {:2.5f}'.format(b))

# Calcular la temperatura para que la cocción sean 5 minutos
tx = 5.0 # 5 minutos en segundos
Tx = m * tx + b

print('Temperatura del huevo para una cocción de 5 [m] : {} [C]'.format(Tx))

In [ ]:
# Grafica de apoyo
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
x = np.linspace(0,6,10)
plt.plot(x,m*x+b)
plt.plot(5,m*5+b,'o')

<a href="./T05_Entrada_Salida_Archivos.ipynb">Entrada, Salida y Archivos</a>

### Ejercicio 6:
Escribir un programa que pregunte al usuario el peso del huevo y crear una lista de temperaturas y una la lista de tiempos de cocción para esas temperaturas; crear un archivo con el nombre "huevo_m_XX", donde XX indica el peso del huevo; escribir en el archivo las listas de temperaturas y tiempos.

In [ ]:
peso_huevo = input('Dame el peso del huevo:')
archivo = open('huevo_m_{}'.format(peso_huevo), 'w')

M = float(peso_huevo)
num1 = M**(2/3) * c * rho**(1/3)
den1 = K * math.pi**2 * (4 * math.pi / 3)**(2/3)

for i in range(4,21):
    t = (num1 /den1) * math.log(0.76 * (i - Tw) / (Ty - Tw))
    archivo.write('{:>3.8f}\t{:>3.8f}\n'.format(i,t/60))
    
archivo.close()

<a href="./T06_Funciones_y_Documentacion.ipynb">Funciones y documentación</a>

<a href="./T07_Excepciones.ipynb">Excepciones</a>

### Proyecto 1:
Organizar los cálculos en funciones que reciban varios tipos de argumentos y que regresen distintos tipos de objetos. La idea es que el programa solicite al usuario la temperatura inicial ($T_o$), el peso del huevo ($M$); El programa debe verificar que $T_o \in [4,20]$ y $M \in [47, 67]$ y que los valores introducidos sean compatibles para realizar los cálculos. Los errores deberán ser manejados con excepciones. Documente sus funciones para que el usuario sepa que es lo que hacen.

<a href="./T08_IterablesMapFilter.ipynb">Iterables, Mapeo y Filtrado</a>

<a href="T09_LambdaExpressions_Reduce.ipynb">Lambda Expressions y Reduce </a>

<a href="T10_Comprehensions.ipynb">Comprehensions</a>


<a href="T11_IteradoresGeneradores.ipynb">Iteradores y generadores</a>

<a href="T12_Decoradores.ipynb">Decoradores</a>

<a href="T13_BibliotecaEstandar.ipynb">Biblioteca estándar</a>

### Proyecto 2:
Re-programe el proyecto 1 usando una forma más *Pythónica*, dando la posibilidad de usar grados Kelvin y Farenheit. 
Bonus: Haga uso de <a href="./BONUS_IPYWidgets.ipynb">IPYWidgets</a> para un mejor manejo del programa.